synapse/handlers/federation.py

```diff
@@ -1667,16 +1667,52 @@ async def on_send_join_request(self, origin: str, pdu: EventBase) -> JsonDict:
         # would introduce the danger of backwards-compatibility problems.
         event.internal_metadata.send_on_behalf_of = origin
 
-        context = await self._handle_new_event(origin, event)
+        # Calculate the event context.
+        context = await self._prep_event(
+            origin, event, state=None, auth_events=None, backfilled=False
+        )
+
+        # Get the current state at the to-be created event.
+        prev_state_ids = await context.get_prev_state_ids()
+
+        # Check if the user is already in the room or invited to the room.
+        user_id = event.state_key
+        prev_member_event_id = prev_state_ids.get((EventTypes.Member, user_id), None)
+        newly_joined = True
+        is_invite = False
+        if prev_member_event_id:
+            prev_member_event = await self.store.get_event(prev_member_event_id)
+            newly_joined = prev_member_event.membership != Membership.JOIN
+            is_invite = prev_member_event.membership == Membership.INVITE
+
+        # We retrieve the room member handler here as to not cause a cyclic dependency
```

Comment:
generally I would take this as a sign that `can_join_without_invite` needs to find somewhere else to live. I appreciate that may be fiddly, though.

Reply:
Fair enough. We do this in a couple of other spots in this file, I think. 😢

Reply:
I'm struggling a bit to find a good spot for this, I was thinking maybe the `RoomHandler`, but that pulls in `RoomMemberHandler` so I think it would have the same issue. Maybe the `EventCreationHandler` or a new handler. 😢 

The other spot we do this seems to be in exchanging 3PIDs.